File: 7SEM/TVP/LAB3/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <Windows.h>
#include <algorithm>

using namespace std;

// Получение индекса переменной в строке
// Функция возвращает индекс первого вхождения символа ch в строке found_variables
int get_index(const string& found_variables, char ch) {
    return found_variables.find(ch);
}

// Проверка корректности входных данных
// Проверяется, содержатся ли символы функции, алфавита, переменных, аксиом и правил в допустимых множествах
bool is_input_correct(const string& function, const string& alphabet, const string& variables, const string& axioms, const vector<string>& left_rules, const vector<string>& right_rules) {
    // Локальная лямбда-функция для проверки строки на допустимые символы
    auto check_symbols = [&](const string& str) {
        for (char character : str) {
            if (alphabet.find(character) == string::npos) {
                cout << "Ошибка: символ '" << character << "' не входит в алфавит!" << endl;
                return false;
            }
        }
        return true;
    };

    // Проверка символов в функции и аксиомах
    if (!check_symbols(function) || !check_symbols(axioms)) return false;

    // Проверка символов в правилах
    for (const string& rule : left_rules) {
        for (char character : rule) {
            if (variables.find(character) == string::npos && alphabet.find(character) == string::npos) {
                cout << "Ошибка: символ '" << character << "' не входит в алфавит или множество переменных!" << endl;
                return false;
            }
        }
    }
    return true;
}

// Чтение и разбор входных данных из файла
// Считывает данные из файла и инициализирует входные переменные
bool read_input(const string& filename, string& function, string& alphabet, string& variables, string& axioms, vector<string>& left_rules, vector<string>& right_rules) {
    ifstream fi(filename);
    if (!fi.is_open()) {
        cout << "Ошибка открытия файла!" << endl;
        return false;
    }

    // Считывание основных строк: функции, алфавита, переменных, аксиом
    getline(fi, function);
    getline(fi, alphabet);
    getline(fi, variables);
    getline(fi, axioms);

    // Лямбда-функция для удаления префиксов, суффиксов и лишних символов
    auto clean_input = [](string& str, const string& prefix, const string& suffix) {
        str.erase(0, prefix.size());
        str.erase(str.size() - suffix.size());
        str.erase(remove(str.begin(), str.end(), ','), str.end());
    };

    // Очистка данных
    clean_input(alphabet, "A={", "};");
    clean_input(variables, "X={", "};");
    clean_input(axioms, "A1={", "};");

    // Считывание правил преобразований
    string line;
    while (getline(fi, line) && line != "}") {
        size_t pos = line.find("->");
        if (pos != string::npos) {
            left_rules.push_back(line.substr(0, pos));
            right_rules.push_back(line.substr(pos + 2));
        }
    }

    // Очистка символов из первой и последней строки правил
    if (!left_rules.empty()) left_rules[0].erase(0, 3);
    if (!right_rules.empty()) right_rules.back().pop_back();

    fi.close();
    return true;
}

// Функция для применения правил
// Заменяет строки в функции на основе заданных правил и записывает результат в файл
void apply_rules(string function, const string& variables, const vector<string>& left_rules, const vector<string>& right_rules) {
    ofstream out("output.txt");
    if (!out.is_open()) {
        cout << "Ошибка открытия файла для записи!" << endl;
        return;
    }

    size_t rule_counter = 0; // Индекс текущего правила
    int return_flag = 0; // Счётчик пропущенных правил
    string found_variables; // Найденные переменные для текущего правила

    // Применение правил, пока не будет пройдена вся таблица
    while (return_flag < left_rules.size()) {
        string reg_ex; // Регулярное выражение для текущего правила
        found_variables.clear();

        // Построение регулярного выражения на основе левой части правила
        for (char ch : left_rules[rule_counter]) {
            if (variables.find(ch) != string::npos) {
                // Если переменная новая, добавляется новый захват
                if (get_index(found_variables, ch) == string::npos) {
                    reg_ex += "(1*)";
                    found_variables += ch;
                } else {
                    reg_ex += "(\\" + to_string(get_index(found_variables, ch) + 1) + ')';
                }
            } else if (ch == '+' || ch == '=') {
                reg_ex += "\\" + string(1, ch); // Обработка специальных символов
            } else {
                reg_ex += ch; // Добавление обычного символа
            }
        }

        regex re(reg_ex); // Создание регулярного выражения
        smatch match; // Для хранения результата сопоставления
        if (regex_search(function, match, re)) {
            // Если правило применимо, записываем результат в файл
            out << "Исходная строка: " << function << endl;
            out << "Применяемое правило: " << left_rules[rule_counter] << " -> " << right_rules[rule_counter] << endl;

            return_flag = 0; // Сбрасываем счётчик пропущенных правил
            string tmp; // Буфер для формирования результата
            vector<string> values(match.size() - 1); // Сохранение значений переменных
            for (size_t i = 1; i < match.size(); ++i) {
                values[i - 1] = match[i];
            }

            // Преобразование по правой части правила
            for (char ch : right_rules[rule_counter]) {
                if (variables.find(ch) != string::npos) {
                    tmp += values[get_index(found_variables, ch)];
                } else {
                    tmp += ch;
                }
            }

            // Замена функции на результат
            function = match.prefix().str() + tmp + match.suffix().str();
            out << "Результат применения правила: " << function << endl << endl;
        } else {
            return_flag++; // Увеличиваем счётчик пропущенных правил
        }

        // Переход к следующему правилу
        rule_counter = (rule_counter + 1) % left_rules.size();
    }

    cout << "Вычисления завершены" << endl;
    out.close();
}

int main() {
    SetConsoleOutputCP(1251); // Установка кодировки консоли для корректного отображения русского текста

    // Инициализация переменных
    string function, alphabet, variables, axioms;
    vector<string> left_rules, right_rules;

    // Чтение данных и выполнение правил, если входные данные корректны
    if (read_input("input.txt", function, alphabet, variables, axioms, left_rules, right_rules) &&
        is_input_correct(function, alphabet, variables, axioms, left_rules, right_rules)) {
        apply_rules(function, variables, left_rules, right_rules);
    }

    return 0;
}

File: 7SEM/TVP/LAB3/main.py
import re

# Получение индекса переменной в строке
def get_index(found_variables, ch):
    return found_variables.find(ch)

# Проверка корректности входных данных
def is_input_correct(function, alphabet, variables, axioms, left_rules, right_rules):
    def check_symbols(s):
        for character in s:
            if character not in alphabet:
                print(f"Ошибка: символ '{character}' не входит в алфавит!")
                return False
        return True

    if not check_symbols(function) or not check_symbols(axioms):
        return False

    for rule in left_rules:
        for character in rule:
            if character not in variables and character not in alphabet:
                print(f"Ошибка: символ '{character}' не входит в алфавит или множество переменных!")
                return False

    return True

# Чтение и разбор входных данных из файла
def read_input(filename):
    try:
        with open(filename, 'r', encoding='utf-8') as fi:
            function = fi.readline().strip()
            alphabet = fi.readline().strip()
            variables = fi.readline().strip()
            axioms = fi.readline().strip()

            # Очистка данных
            def clean_input(s, prefix, suffix):
                return s[len(prefix):-len(suffix)].replace(',', '')

            alphabet = clean_input(alphabet, "A={", "};")
            variables = clean_input(variables, "X={", "};")
            axioms = clean_input(axioms, "A1={", "};")

            # Считывание правил преобразований
            left_rules, right_rules = [], []
            for line in fi:
                line = line.strip()
                if line == "}":
                    break
                if "->" in line:
                    pos = line.find("->")
                    left_rules.append(line[:pos])
                    right_rules.append(line[pos + 2:])

            if left_rules:
                left_rules[0] = left_rules[0][3:]
            if right_rules:
                right_rules[-1] = right_rules[-1][:-1]

            return function, alphabet, variables, axioms, left_rules, right_rules
    except FileNotFoundError:
        print("Ошибка открытия файла!")
        return None

# Функция для применения правил
def apply_rules(function, variables, left_rules, right_rules):
    try:
        with open("D:/GIT/VuzUC/7SEM/TVP/LAB3/output.txt", "w", encoding="utf-8") as out:
            rule_counter = 0
            return_flag = 0
            found_variables = ""

            while return_flag < len(left_rules):
                reg_ex = ""
                found_variables = ""

                for ch in left_rules[rule_counter]:
                    if ch in variables:
                        if ch not in found_variables:
                            reg_ex += "(1*)"
                            found_variables += ch
                        else:
                            reg_ex += f"(\\{get_index(found_variables, ch) + 1})"
                    elif ch in "+=":
                        reg_ex += re.escape(ch)
                    else:
                        reg_ex += ch

                match = re.search(reg_ex, function)
                if match:
                    out.write(f"Исходная строка: {function}\n")
                    out.write(f"Применяемое правило: {left_rules[rule_counter]} -> {right_rules[rule_counter]}\n")

                    return_flag = 0
                    tmp = ""
                    values = [match.group(i) for i in range(1, len(match.groups()) + 1)]

                    for ch in right_rules[rule_counter]:
                        if ch in variables:
                            tmp += values[get_index(found_variables, ch)]
                        else:
                            tmp += ch

                    function = match.string[:match.start()] + tmp + match.string[match.end():]
                    out.write(f"Результат применения правила: {function}\n\n")
                else:
                    return_flag += 1

                rule_counter = (rule_counter + 1) % len(left_rules)

            print("Вычисления завершены")
    except IOError:
        print("Ошибка открытия файла для записи!")

# Главная функция
if __name__ == "__main__":
    # Инициализация переменных
    data = read_input("D:/GIT/VuzUC/7SEM/TVP/LAB3/input.txt")
    if data:
        function, alphabet, variables, axioms, left_rules, right_rules = data
        if is_input_correct(function, alphabet, variables, axioms, left_rules, right_rules):
            apply_rules(function, variables, left_rules, right_rules)
